Add incremental SMT2 sort checked construction of bit vector predicate application and application of core theory functions

thomasspriggs opened the pull request:
Add incremental SMT2 sort checked construction of bit vector predicate application and application of core theory functions.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

TGWDB commented on src/solvers/smt2_incremental/smt_commands.cpp:
```diff
@@ -198,3 +198,51 @@ void smt_commandt::accept(smt_command_const_downcast_visitort &&visitor) const
+  return _identifier.identifier();
+}
+
+smt_sortt smt_command_functiont::return_sort(
+  const std::vector<smt_termt> &arguments) const
+{
+  return _identifier.get_sort();
+}
```
It's not clear here why the `return_sort` function needs the `arguments` parameter.

thomasspriggs replied:
It doesn't in this particular case. But it might need it for things like "bitvector bitwise or" or "ite" from the CORE theory for example.

thomasspriggs commented:
This PR relates to https://github.com/diffblue/cbmc/issues/6134

TGWDB commented on src/solvers/smt2_incremental/smt_bitvector_theory.cpp:
```diff
@@ -0,0 +1,42 @@
+  const smt_termt &left,
+  const smt_termt &right)
+{
+  const auto left_sort = left.get_sort().cast<smt_bit_vector_sortt>();
+  INVARIANT(left_sort, "Left operand must have bitvector sort.");
+  const auto right_sort = right.get_sort().cast<smt_bit_vector_sortt>();
+  INVARIANT(right_sort, "Right operand must have bitvector sort.");
+  INVARIANT(
+    left_sort->bit_width() == right_sort->bit_width(),
+    "Left and right operands must have the same bit width.");
+}
```
Do we need to check they are both signed/unsigned?

thomasspriggs replied:
Bit vectors in SMT2 bitvector theory do not have signedness as a property. Therefore this information is not available to this factory enforcing these invariants. Good ways to check use of the correct kind of operator might be worth us discussing over slack.

TGWDB commented on src/solvers/smt2_incremental/smt_bitvector_theory.cpp:
```diff
@@ -0,0 +1,42 @@
+    validate_bitvector_predicate_arguments(left, right);                       \
+  }                                                                            \
+                                                                               \
+  smt_sortt smt_bit_vector_theoryt::the_name##t::return_sort(                  \
+    const smt_termt &, const smt_termt &)                                      \
+  {                                                                            \
+    return smt_bool_sortt{};                                                   \
+  }                                                                            \
```
Do we need the arguments here?

thomasspriggs replied:
No. Which is why they are unnamed. It is more straightforward to leaved them unused than it is to make the template code pass them only in some circumstances.

TGWDB reviewed (commented):
There are a lot of tests for different operators, but they all use the same factory code. Is there significant value in having them all?

thomasspriggs committed:
Add sort casting

For checking parameterised sorts.

thomasspriggs committed:
Add parameterised function application factory

So that function specific INVARIANTS can be checked.

thomasspriggs committed:
Add bitvector theory predicates

thomasspriggs committed:
Refactor `smt_not_term` into core theory

For consistency with how the functions for the bit vector theory are
defined. Note that the factory is called `make_not` rather than `not`,
because not is a keyword in C++ which is equivalent to the `!` operator.

thomasspriggs committed:
Make `smt_function_application_termt` constructor private

To prevent construction of function application without checking that
the given arguments are valid for the given function.

thomasspriggs committed:
Add SMT2 core theory "distinct"

thomasspriggs committed:
Add core theory "if then else" operation

thomasspriggs committed:
Add core theory "implies" operation

thomasspriggs committed:
Add SMT2 incremental core theory logical operators

thomasspriggs committed:
Add `equal` to smt core theory

chrisr-diffblue commented on src/solvers/smt2_incremental/smt_commands.cpp:
```diff
@@ -198,3 +198,51 @@ void smt_commandt::accept(smt_command_const_downcast_visitort &&visitor) const
+{
+  INVARIANT(
+    parameter_sorts.size() == arguments.size(),
+    "Number of parameters and number of arguments must the the same.");
```
"must the the same" => "must be the same"

thomasspriggs replied:
Done.

chrisr-diffblue commented on unit/solvers/smt2_incremental/smt_commands.cpp:
```diff
@@ -94,3 +95,15 @@ TEST_CASE("smt_set_option_commandt getter", "[core][smt2_incremental]")
   CHECK(smt_set_option_commandt{models_off}.option() == models_off);
 }
+
+TEST_CASE("SMT2 function application factory tests", "[core][smt2_incremental]")
```
No `CHECK`s in this test? Whats its purpose?

thomasspriggs replied:
The checks were missing because I got distracted adding other functionality to this PR, before I finished writing this test. Thanks for spotting this oversight. I have now added some checks.
